### PR TITLE
Add reader for Silixa's HDF5 iDAS files

### PR DESCRIPTION
**Pull request by fablindner**

### CODE REVIEW

**Comment by emolch** [comment]
Hi @fablindner, thanks a lot for sharing this reader!

Could you please delay the import of h5py into the `iload` and `get_meta` functions. This would save us from having a hard dependency on h5py. Maybe you can pack the import into a try/except and give a meaningful error message with a pointer on how to install h5py.

Would it be possible to make the `detect` function a bit more specific? I am a bit concerned about false-positives as there are many other HDF5-based files out there.

**Comment by fablindner** [comment]
Hi @emolch , thanks for checking and the instructions. I added the suggested try/except statement for the h5py import. 

Regarding the detect function, I unfortunately could not come up with a more specific identifier. I checked against other HDF5 files (not containing DAS data), but at the moment I cannot see how to discriminate them, sorry.

File: src/io/hdf5_idas.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    return meta
+
+
+def iload(filename, load_data=True):
+
+    # prevent hard dependency on h5py
+    try:
+        import h5py
+    except Exception as e:
+        print(e)
```
Comment (miili) [inline src/io/hdf5_idas.py]: remove print. Please use `raise ImportError(...) from e`. See https://docs.python.org/3/tutorial/errors.html#exception-chaining

File: src/io/hdf5_idas.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        # get the sampling rate, starttime, number of samples in space and time
+        deltat = 1. / f['Acquisition/Raw[0]'].attrs['OutputDataRate']
+        tmin = datetime.strptime(
+            f['Acquisition/Raw[0]/RawData'].attrs['PartStartTime'].decode('ascii'),
+            '%Y-%m-%dT%H:%M:%S.%f+00:00').timestamp()
+        nchan = f['Acquisition/Raw[0]'].attrs['NumberOfLoci']
+        try:
+            # usually this should be here?
+            nsamp = f['Acquisition/Raw[0]/RawDataTime'].attrs['Count'] 
+        except:
```
Comment (miili) [inline src/io/hdf5_idas.py]: Please use an explicit Exception. `KeyError`?

File: src/io/hdf5_idas.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+            # but sometimes it is here?
+            nsamp = f['Acquisition/Raw[0]/RawData'].attrs['Count'] 
+
+    for icha in range(nchan):
+        
+        assert icha < 99999
```
Comment (miili) [inline src/io/hdf5_idas.py]: `assert` should only be used in tests. Please use `raise ValueError` instead. The other reader should do the same 🙈

File: src/io/hdf5_idas.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    except Exception as e:
+        print(e)
+        raise ImportError("Please install 'h5py' to proceed, e.g. by running 'pip install h5py'")
+
+
+    with h5py.File(filename, "r") as f:
+        # get the meta data
+        meta = get_meta(f)
+        # get the actual time series if load_data
+        data = f['Acquisition/Raw[0]/RawData'][:].copy() if load_data else None
```
Comment (miili) [inline src/io/hdf5_idas.py]: Is the `copy` needed? Or can we get away with the view?

File: src/io/hdf5_idas.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        # get the sampling rate, starttime, number of samples in space and time
+        deltat = 1. / f['Acquisition/Raw[0]'].attrs['OutputDataRate']
+        tmin = datetime.strptime(
+            f['Acquisition/Raw[0]/RawData'].attrs['PartStartTime'].decode('ascii'),
+            '%Y-%m-%dT%H:%M:%S.%f+00:00').timestamp()
```
Comment (miili) [inline src/io/hdf5_idas.py]: Use `datetime.fromisotimestamp` instead of `strptime`

File: src/io/hdf5_idas.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                  'Acquisition/Custom/SystemSettings',
+                  'Acquisition/Custom/SystemInformation/GPS',
+                  'Acquisition/Custom/UserSettings',
+                  'Acquisition/Raw[0]']:
+        try:
+            field_keys = h5file[field].attrs.keys()
+            for val in val_list:
+                if val in field_keys:
+                    meta[key_list[val_list.index(val)]] = f[field].attrs[val]
+        except:
```
Comment (miili) [inline src/io/hdf5_idas.py:66]: Please use an explicit exception. `KeyError`?

**Review by miili — CHANGES_REQUESTED** [review]
Thanks for this reader! How well does it perform in the snuffler UI?

File: src/io/hdf5_idas.py (hunk 1 above)
Comment (fablindner) [reply src/io/hdf5_idas.py]: Done

File: src/io/hdf5_idas.py (hunk 2 above)
Comment (fablindner) [reply src/io/hdf5_idas.py]: I removed this try-except statement and only kept line 103, which should work for files with complete header information.

File: src/io/hdf5_idas.py (hunk 3 above)
Comment (fablindner) [reply src/io/hdf5_idas.py]: Removed this completely as it can also be larger than 99999?

File: src/io/hdf5_idas.py (hunk 4 above)
Comment (fablindner) [reply src/io/hdf5_idas.py]: No, not need. Removed it.

File: src/io/hdf5_idas.py (hunk 5 above)
Comment (fablindner) [reply src/io/hdf5_idas.py]: Done

File: src/io/hdf5_idas.py (hunk 6 above)
Comment (fablindner) [reply src/io/hdf5_idas.py:66]: I removed this try-except statement, which should work for files with complete header information.

**Comment by fablindner** [comment]
Thanks for the suggestions for improvement, I tried to implement them...sorry for the late reply! The reader works generally well. However, one problem that I encounter: when calling snuffler on some files, the preview of the available data does not have the correct timestamps (shifted by one hour). But when zooming in to see the actual wiggles (or the waterfall) plot, the data jump to the correct position in time. I observed a similar problem with the tdms_idas reader, so there might be an issue somewhere else and not in the reader functions.

File: src/io/hdf5_idas.py (hunk 3 above)
Comment (miili) [reply src/io/hdf5_idas.py]: The miniSEED standard which we leverage only allows for 5 characters, so we have to limit it to 5 chars.

File: src/io/hdf5_idas.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+    with h5py.File(filename, "r") as f:
+        # get the meta data
+        meta = get_meta(f)
+        # get the actual time series if load_data
+        data = f['Acquisition/Raw[0]/RawData'][:] if load_data else None
+        # get the sampling rate, starttime, number of samples in space and time
+        deltat = 1. / f['Acquisition/Raw[0]'].attrs['OutputDataRate']
+        tmin = datetime.fromisoformat(
+            f['Acquisition/Raw[0]/RawData'].attrs['PartStartTime'].decode('ascii')).timestamp()
```
Comment (miili) [inline src/io/hdf5_idas.py:96]: Is this always UTC?

to enforce UTC you can either
1. append `+00:00` to the iso string
2. use `datetime.replace(tzinfo=timezone.utc)`

**Comment by miili** [comment]
> Thanks for the suggestions for improvement, I tried to implement them...sorry for the late reply! The reader works generally well. However, one problem that I encounter: when calling snuffler on some files, the preview of the available data does not have the correct timestamps (shifted by one hour). But when zooming in to see the actual wiggles (or the waterfall) plot, the data jump to the correct position in time. I observed a similar problem with the tdms_idas reader, so there might be an issue somewhere else and not in the reader functions.

Interesting observation, I will have an eye on that with incoming data sets. Sorry for my late reply, as well. Let's get this through!

File: src/io/hdf5_idas.py (hunk 3 above)
Comment (emolch) [reply src/io/hdf5_idas.py]: Pyrocko and Snuffler can handle more than 5 characters, so it is ok to remove the limit. Only on an attempt to save the traces as miniSEED we will get an error, but that is ok imo.

File: src/io/hdf5_idas.py (hunk 7 above)
Comment (fablindner) [reply src/io/hdf5_idas.py:96]: The raw time string already contains the `+00:00` part.

File: src/io/hdf5_idas.py (hunk 3 above)
Comment (fablindner) [reply src/io/hdf5_idas.py]: Yes, saw that too, that more than 5 characters are okay for the pyrocko trace. So I can leave it like this?

**Comment by miili** [comment]
Merged on gitea, @fablindner thanks again!

**Pull request merged**